packages/SwingSet/src/kernel/kernel.js

```diff
@@ -497,12 +501,22 @@ export default function buildKernel(
       } else {
         throw Error(`unable to process message.type ${message.type}`);
       }
-      if (deliveryProblem) {
-        abortCrank();
-        const { vatID, problem } = deliveryProblem;
-        terminateVat(vatID, problem);
-        kdebug(`vat abnormally terminated: ${problem}`);
-      } else {
+      let didAbort = false;
+      if (terminationTrigger) {
+        const {
+          vatID,
+          shouldAbortCrank,
+          shouldReject,
+          info,
+        } = terminationTrigger;
+        if (shouldAbortCrank) {
+          abortCrank();
+          didAbort = true;
+        }
+        terminateVat(vatID, shouldReject, info);
+        kdebug(`vat terminated: ${JSON.stringify(info)}`);
+      }
+      if (!didAbort) {
```

Comment:
I have this weak feeling that maybe `purgeDeadKernelPromises` and/or `saveStats` should still be called if we terminated a vat but didn't abort its crank side-effects.. not sure, could you take a second look?

Reply:
Yes, your intuition is correct, and that is what the code there does.  Am I missing something?